packages/expo-constants/android/src/main/java/expo/modules/constants/ConstantsService.kt

```diff
@@ -0,0 +1,142 @@
+
+class ConstantsService(private val context: Context) : InternalModule, ConstantsInterface {
+  private var statusBarHeight = 0
+  private val sessionId = UUID.randomUUID().toString()
+  private val exponentInstallationId: ExponentInstallationId = ExponentInstallationId(context)
+
+  enum class ExecutionEnvironment(val string: String) {
+    BARE("bare"),
+    STANDALONE("standalone"),
+    STORE_CLIENT("storeClient");
+  }
+
+  override fun getExportedInterfaces(): List<Class<*>> =
+    listOf(ConstantsInterface::class.java as Class<*>)
+
+  override fun getConstants(): Map<String, Any> {
+    val platform: Map<String, Any> = mapOf("android" to hashMapOf<String, Any>(/* empty */))
+
+    val constants: MutableMap<String, Any> = hashMapOf(
+      "sessionId" to sessionId,
+      "executionEnvironment" to ExecutionEnvironment.BARE.string,
+      "statusBarHeight" to statusBarHeight,
+      "deviceYearClass" to deviceYearClass,
+      "deviceName" to deviceName,
+      "isDevice" to isDevice,
+      "systemFonts" to systemFonts,
+      "systemVersion" to systemVersion,
+      "installationId" to getOrCreateInstallationId,
+      "manifest" to appConfig!!,
+      "platform" to platform,
+    )
+
+    try {
+      val pInfo = context.packageManager.getPackageInfo(context.packageName, 0)
+      constants["nativeAppVersion"] = pInfo.versionName
+
+      val versionCode = getLongVersionCode(pInfo).toInt()
+      constants["nativeBuildVersion"] = versionCode.toString()
+    } catch (e: PackageManager.NameNotFoundException) {
+      Log.e(TAG, "Exception: ", e)
+    }
+
+    return constants
+  }
+
+  // Just use package name in vanilla React Native apps.
+  override fun getAppScopeKey(): String = context.packageName
+
+  override fun getAppOwnership() = "guest"
+
+  override fun getDeviceName(): String = Build.MODEL
+
+  override fun getDeviceYearClass() = YearClass.get(context)
+
+  override fun getIsDevice() = !isRunningOnGenymotion && !isRunningOnStockEmulator
+
+  override fun getStatusBarHeight() = statusBarHeight
+
+  override fun getSystemVersion(): String = Build.VERSION.RELEASE
+
+  val getOrCreateInstallationId: String
+    get() = exponentInstallationId.orCreateUUID
+
+  // From https://github.com/dabit3/react-native-fonts
+  override fun getSystemFonts() = listOf(
+    "normal",
+    "notoserif",
+    "sans-serif",
+    "sans-serif-light",
+    "sans-serif-thin",
+    "sans-serif-condensed",
+    "sans-serif-medium",
+    "serif",
+    "Roboto",
+    "monospace"
+  )
+
+  private val appConfig: String?
+    get() {
+      try {
+        context.assets.open(CONFIG_FILE_NAME).use {
+          stream ->
+          return IOUtils.toString(stream, StandardCharsets.UTF_8)
+        }
+      } catch (e: FileNotFoundException) {
+        // do nothing, expected in managed apps
+      } catch (e: Exception) {
+        Log.e(TAG, "Error reading embedded app config", e)
+      }
+      return null
+    }
```

Comment:
I was not able to find better alternative for java's [`try-with-resources`](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) syntax than `use`. Suggestions here are especially welcome.